src/Lecture2.hs

```diff
@@ -74,7 +87,15 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt index list
+  | index < 0 = (Nothing, list)
+  | otherwise = go 0 list []
+  where
+    go _ [] before = (Nothing, reverse before)
+    go i (x:xs) before
+      | i == index = (Just x, reverse before ++ xs)
+      | otherwise = go (i + 1) xs (x : before)
```

Comment:
Should I construct the list with before ++ [x]? I thought that using ++ is also not very efficient.

Reply:
Indeed, using `++ [x]` is considered an anti-pattern and quite inefficient.

The ideal solution uses `:` to construct the list and doesn't `reverse` the list in the end

Reply:
Implemented a variant without reverse and concat:
```
removeAt :: Int -> [a] -> (Maybe a, [a])
removeAt index list
  | index < 0 = (Nothing, list)
  | otherwise = go 0 list id
  where
    go :: Int -> [a] -> ([a] -> [a]) -> (Maybe a, [a])
    go _ [] buildPrev = (Nothing, buildPrev [])
    go i (x : xs) buildPrev
      | i == index = (Just x, buildPrev xs)
      | otherwise = go (i + 1) xs (buildPrev . (x :))
```

Reply:
Using the "difference list" approach is an elegant solution here! 👏🏻 

Alternatively, you could use the recursion like this:

```haskell
removeAt :: Int -> [a] -> (Maybe a, [a])
removeAt index list
  | index < 0 = (Nothing, list)
  | otherwise = go 0 list
  where
    go :: Int -> [a] -> (Maybe a, [a])
    go _ [] = (Nothing, [])
    go i (x : xs)
      | i == index = (Just x, xs)
      | otherwise = 
          let (val, ys) = go (i + 1) xs
          in (val, x : ys)
```

Reply:
Ok. Interesting solution! Thank you!